### ISSUE TITLE
step3_reduceChronos runs very long without any logging message that informs the progress

### ISSUE THREAD

**Issue by kazurayam** [kazurayam/VisualInspectionOfExcelAndPDF#14]
I exected "Test Cases/NISA/main_Chronos",
the "step3_reduceChronos" started and ran long, long minutes.

<img width="1552" alt="スクリーンショット 2023-04-29 21 48 41" src="https://user-images.githubusercontent.com/4467417/235303393-2f942423-fa89-4a53-af7d-87feb72e8412.png">

The problem is that it emits no logging messages. Therefore I can not see (guess) what's going on.

I want to let "step3_reduceChronos" to emit something informative.

**Comment by kazurayam** [comment]
I edited `Include/confing/log.properties` file to have a line

```
logging.level.com.kazurayam.materialstore.base.reduce.DiffingMPGProcessor=DEBUG
```

then I removed the `store` directory and the `store-backup` directory. I ran `Test Cases/NISA/main_Chronos`, I saw in the console

```
2023-04-30 08:56:12.225 INFO  c.k.k.c.keyword.builtin.CommentKeyword   - step3_reduceChronos started
2023-04-30 08:56:12.288 INFO  c.k.k.c.keyword.builtin.CommentKeyword   - going to select the current MaterialList
2023-04-30 08:56:12.294 INFO  c.k.k.c.keyword.builtin.CommentKeyword   - will compare the current JobTimestamp against the baseline specified as parameter
2023-04-30 08:56:13.588 INFO  c.k.k.c.keyword.builtin.CommentKeyword   - going to call inspector.reduceAndSort(reduced)
2023-04-30 08:56:13.589 DEBUG c.k.m.base.reduce.DiffingMPGProcessor    - #process mpg.getCountTotal()=44
2023-04-30 08:56:13.592 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"0f58b2d68b0e043644b8b1436f4814f87cf8e459","fileType":"csv","metadata":{"seq":{"key":"seq","value":"1","matchedByAster":true}, "sheet_index":{"key":"sheet_index","value":"0","matchedByAster":true}, "sheet_name":{"key":"sheet_name","value":"指定インデックス投資信託","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/23.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"1\", \"sheet_index\":\"0\", \"sheet_name\":\"指定インデックス投資信託\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/23.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/0f58b2d68b0e043644b8b1436f4814f87cf8e459.csv","diffability":"AS_TEXT"}
2023-04-30 08:56:15.221 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"a057919e32886466213e5393068d27f39bd48f5f","fileType":"csv","metadata":{"seq":{"key":"seq","value":"1","matchedByAster":true}, "sheet_index":{"key":"sheet_index","value":"1","matchedByAster":true}, "sheet_name":{"key":"sheet_name","value":"指定インデックス投資信託以外の投資信託（アクティブ運用投信等）","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/23.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"1\", \"sheet_index\":\"1\", \"sheet_name\":\"指定インデックス投資信託以外の投資信託（アクティブ運用投信等）\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/23.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/a057919e32886466213e5393068d27f39bd48f5f.csv","diffability":"AS_TEXT"}
2023-04-30 08:56:15.503 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"c97fb9280a513fd9bf90904ed36e756bae5873c8","fileType":"csv","metadata":{"seq":{"key":"seq","value":"1","matchedByAster":true}, "sheet_index":{"key":"sheet_index","value":"2","matchedByAster":true}, "sheet_name":{"key":"sheet_name","value":"上場株式投資信託（ETF）","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/23.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"1\", \"sheet_index\":\"2\", \"sheet_name\":\"上場株式投資信託（ETF）\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/23.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/c97fb9280a513fd9bf90904ed36e756bae5873c8.csv","diffability":"AS_TEXT"}
2023-04-30 08:56:15.790 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"cd1befebe8336e3837aeffd2856220a4e7b7e4ab","fileType":"xlsx","metadata":{"seq":{"key":"seq","value":"1","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/23.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"1\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/23.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","metadataURL":"https://www.fsa.go.jp/policy/nisa/20170614-2/23.xlsx","relativeUrl":"NISA_Chronos/20230430_085419/objects/cd1befebe8336e3837aeffd2856220a4e7b7e4ab.xlsx","diffability":"UNABLE"}
2023-04-30 08:56:15.833 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"3a98c4ba471f11462d06a4c94ef4daa4010a466a","fileType":"png","metadata":{"page":{"key":"page","value":"1","matchedByAster":true}, "seq":{"key":"seq","value":"5","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/24.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"page\":\"1\", \"seq\":\"5\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/24.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/3a98c4ba471f11462d06a4c94ef4daa4010a466a.png","diffability":"AS_IMAGE"}
2023-04-30 08:59:10.660 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"e52e94169bdbd567a0d7d26cfa36a1bb93da14d9","fileType":"png","metadata":{"page":{"key":"page","value":"2","matchedByAster":true}, "seq":{"key":"seq","value":"5","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/24.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"page\":\"2\", \"seq\":\"5\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/24.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/e52e94169bdbd567a0d7d26cfa36a1bb93da14d9.png","diffability":"AS_IMAGE"}
2023-04-30 09:01:37.391 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"be9a8d60edba802ff992c624882d19aec4a8fd94","fileType":"png","metadata":{"page":{"key":"page","value":"3","matchedByAster":true}, "seq":{"key":"seq","value":"5","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/24.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"page\":\"3\", \"seq\":\"5\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/24.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/be9a8d60edba802ff992c624882d19aec4a8fd94.png","diffability":"AS_IMAGE"}
2023-04-30 09:03:47.191 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"bef67e35b6b06faf9a6652bbbbd9d0e803f3e452","fileType":"png","metadata":{"page":{"key":"page","value":"4","matchedByAster":true}, "seq":{"key":"seq","value":"5","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/24.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"page\":\"4\", \"seq\":\"5\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/24.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/bef67e35b6b06faf9a6652bbbbd9d0e803f3e452.png","diffability":"AS_IMAGE"}
2023-04-30 09:05:35.587 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"ecbb466846addc730a8a48edacff58dcf105434e","fileType":"png","metadata":{"page":{"key":"page","value":"5","matchedByAster":true}, "seq":{"key":"seq","value":"5","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/24.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"page\":\"5\", \"seq\":\"5\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/24.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/ecbb466846addc730a8a48edacff58dcf105434e.png","diffability":"AS_IMAGE"}
2023-04-30 09:07:40.465 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"17c65e5a9a1478ceaf46a787bcce0e9d9b52c2ea","fileType":"pdf","metadata":{"seq":{"key":"seq","value":"5","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/24.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"5\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/24.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","metadataURL":"https://www.fsa.go.jp/policy/nisa/20170614-2/24.pdf","relativeUrl":"NISA_Chronos/20230430_085419/objects/17c65e5a9a1478ceaf46a787bcce0e9d9b52c2ea.pdf","diffability":"UNABLE"}
2023-04-30 09:07:40.498 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"fd852582e781a4571d56e22b5a6bc0ddcd231d6b","fileType":"csv","metadata":{"seq":{"key":"seq","value":"2","matchedByAster":true}, "sheet_index":{"key":"sheet_index","value":"0","matchedByAster":true}, "sheet_name":{"key":"sheet_name","value":"指定インデックス投資信託","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/25.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"2\", \"sheet_index\":\"0\", \"sheet_name\":\"指定インデックス投資信託\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/25.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/fd852582e781a4571d56e22b5a6bc0ddcd231d6b.csv","diffability":"AS_TEXT"}
2023-04-30 09:07:40.599 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"b5e992288255eaab804fb0aee64ed3ef2c6f12aa","fileType":"csv","metadata":{"seq":{"key":"seq","value":"2","matchedByAster":true}, "sheet_index":{"key":"sheet_index","value":"1","matchedByAster":true}, "sheet_name":{"key":"sheet_name","value":"指定インデックス投資信託以外の投資信託（アクティブ運用投信等）","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/25.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"2\", \"sheet_index\":\"1\", \"sheet_name\":\"指定インデックス投資信託以外の投資信託（アクティブ運用投信等）\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/25.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/b5e992288255eaab804fb0aee64ed3ef2c6f12aa.csv","diffability":"AS_TEXT"}
2023-04-30 09:07:40.728 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"66719fac859324ed566f6598884379c2fcbf70ef","fileType":"csv","metadata":{"seq":{"key":"seq","value":"2","matchedByAster":true}, "sheet_index":{"key":"sheet_index","value":"2","matchedByAster":true}, "sheet_name":{"key":"sheet_name","value":"上場株式投資信託（ETF）","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/25.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"2\", \"sheet_index\":\"2\", \"sheet_name\":\"上場株式投資信託（ETF）\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/25.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/66719fac859324ed566f6598884379c2fcbf70ef.csv","diffability":"AS_TEXT"}
2023-04-30 09:07:40.826 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"f19feb737fd34872633859f36b786d1459febd9e","fileType":"xlsx","metadata":{"seq":{"key":"seq","value":"2","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/25.xlsx","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"seq\":\"2\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/25.xlsx\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","metadataURL":"https://www.fsa.go.jp/policy/nisa/20170614-2/25.xlsx","relativeUrl":"NISA_Chronos/20230430_085419/objects/f19feb737fd34872633859f36b786d1459febd9e.xlsx","diffability":"UNABLE"}
2023-04-30 09:07:40.836 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"163e076b29fb0dc4f5ed433b41ab8b695ea30d5c","fileType":"png","metadata":{"page":{"key":"page","value":"1","matchedByAster":true}, "seq":{"key":"seq","value":"6","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/26.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"page\":\"1\", \"seq\":\"6\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/26.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/163e076b29fb0dc4f5ed433b41ab8b695ea30d5c.png","diffability":"AS_IMAGE"}
2023-04-30 09:10:47.750 WARN  c.k.m.base.reduce.DiffingMPGProcessor    - #stuffDiffByDiffer left Material was NULL_OBJECT. right={"jobName":"NISA_Chronos","jobTimestamp":"20230430_085419","id":"b6b387947bb407980a632c38a3c46f7b8dada4a0","fileType":"png","metadata":{"page":{"key":"page","value":"2","matchedByAster":true}, "seq":{"key":"seq","value":"6","matchedByAster":true}, "URL.host":{"key":"URL.host","value":"www.fsa.go.jp","matchedByAster":true}, "URL.path":{"key":"URL.path","value":"/policy/nisa/20170614-2/26.pdf","matchedByAster":true}, "URL.port":{"key":"URL.port","value":"80","matchedByAster":true}, "URL.protocol":{"key":"URL.protocol","value":"https","matchedByAster":true}},"identification":"{\"page\":\"2\", \"seq\":\"6\", \"URL.host\":\"www.fsa.go.jp\", \"URL.path\":\"/policy/nisa/20170614-2/26.pdf\", \"URL.port\":\"80\", \"URL.protocol\":\"https\"}","relativeUrl":"NISA_Chronos/20230430_085419/objects/b6b387947bb407980a632c38a3c46f7b8dada4a0.png","diffability":"AS_IMAGE"}
...
```

This message told me that the DiffiingMPGProcessor#stuffDiffByDiffer() is actually processing the entries of MaterialProductGroup, but each took long --- 2 minuites. Why? Possbily because there is no "left" Material found.

**Comment by kazurayam** [comment]
If no "left" Materials found, it does not make sense to try stuffing diff. In that case, the program should skip the heavy workload. But in fact the program seems to be doing a heavy job. 

I should review the code of DiffingMPGProcessor.

**Comment by kazurayam** [comment]
fixed by the materialstore-0.16.2

**Comment by kazurayam** [comment]
done at [v0.14.2-SNAPSHOT](https://github.com/kazurayam/VisualInspectionOfExcelAndPDF/releases/tag/0.14.2-SNAPSHOT)
